### ISSUE TITLE
Autotune times an unparallelized serial baseline on GPU: multi-hour uninterruptible dispatches that monopolize the device

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#532]
Found during the ROCm leg of #476 (minix, gfx1151/RDNA3.5, ROCm 7.14, WSL2, tree at `bd075cd5`).

`Autotune.tune` hangs on some workloads before producing any candidate at all:

- constant ~3-thread CPU spin, indefinitely (killed at >65 min, and again by a 50-min watchdog)
- **zero** `autotune_log` lines after `tune_placements: arm A (default placements) search:` (the log flushes per candidate, so this is not buffering)
- with `--ocannl_output_debug_files_in_build_directory=true`, **zero** debug artifacts written — so it never reaches rendering or hiprtc, not even for the serial baseline

So the wedge is in the middle end (lowering / scheduling / digesting) of the GPU arm, before codegen.

**Affected / not affected, same tree:**

| cell | outcome |
|---|---|
| `mlp_wide` hip/tuned | wedges (reproduced 3×) |
| `cifar_conv` hip/tuned | wedges |
| `mlp_small` hip/tuned | completes, ~40 s |
| `mlp_wide` hip/default | fine, 2.28 ms/step |
| `mlp_wide` cc/tuned | searches normally (`F_sketch[mma-cpu 128x128x128]` timed at 34.7 ms) |

Repro:

```bash
cd benchmarks && BENCH_FIXTURE=fixtures/mlp_wide.safetensors BENCH_TUNE=1 \
  ../_build/default/benchmarks/runners/ocannl/bench_mlp.exe --ocannl_backend=hip \
  --ocannl_autotune_log=true --ocannl_autotune_cache_dir=/tmp/scratch_cache
```

**Possibly a regression.** `benchmarks/report-hip.md`'s previous revision has `mlp_wide hip/tuned` completing at `8436e362` with `compile_s` 111.78 s — but that was recorded on native Windows, so "regression since the gh-502 seeding wave" and "WSL-specific" are both live hypotheses and I did not bisect. A `git bisect` of the hip/tuned cell against `8436e362`, plus one native-Windows run of the current tree, would separate them.

Worked around for now: `("mlp_wide", "hip", "tuned")` added to `SKIP_CELLS` in `benchmarks/orchestrate.py`.

**Comment by lukstafi** [comment]
## The gh-502 hypothesis is refuted for `mlp_wide`, by inspection

Code-level bisect rather than a `git bisect` — no HIP hardware to hand — but it is decisive for the named hypothesis, because **both halves of the gh-502 seeding wave are inert on a conv-free workload**:

- **`a203cebf`** (Autotune: seed conv sketches on strided rows) changes only `conv_seed_params`, which is reached through `detect_conv` (`arrayjit/lib/autotune.ml:1418`). `mlp_wide` is a pure relu MLP with no convolution index patterns, so `detect_conv` returns `None` and the function returns before any changed line executes.
- **`81fa1b7a`** (Schedule.Stage: compact strided tile parts) triggers only for "a single-term tile part with coefficient `c > 1`" — a strided window. MLP GEMM staging is unit-stride, so compaction never fires.

Neither changed line can run on this cell. Whatever wedges `mlp_wide` predates the seeding wave or is unrelated to it.

**Caveat: this covers `mlp_wide` only.** `cifar_conv` — the other wedging cell — *is* a conv workload, so gh-502 remains a live hypothesis there. They may be two bugs sharing a symptom, and are worth testing separately rather than as one.

## It also does not belong with #521

Opposite phase. #521's candidates are all enumerated and *then* fail at candidate compile — that is exactly how its blocker tables exist. Here nothing is enumerated at all: zero `autotune_log` lines after the arm-A banner, zero debug artifacts even with `output_debug_files_in_build_directory=true`. A fix for #521 would not touch this, and vice versa.

## What the evidence does point at

Two discriminators from the issue's own table:

- **Size-dependent**: `mlp_small` completes in ~40 s, `mlp_wide` wedges. Same workload family, same code path, differing only in extents.
- **Backend-dependent**: `cc` and `metal` both complete `mlp_wide`; only HIP wedges.

Together those point at middle-end work parameterized by the backend's `hardware_limits` / `mma_format_tiles` — gfx1151's 16x16x16 against Metal's 8x8x8 — with a cost that grows with extents. Seed enumeration or digesting driven by those numbers is the region to look in; conv seeding is not.

## Cheapest next step on minix — not a bisect

Run `mlp_wide` hip/tuned at `8436e362` **under WSL**. The old passing measurement (`compile_s` 111.78 s) was recorded on native Windows, so that single run separates the issue's two stated hypotheses — "WSL-specific" vs "tree regression" — before anyone spends time on a bisect. And since gh-502 is refuted for this cell, a bisect would need a range much wider than lukstafi/ocannl-staging#227, so it is worth knowing which hypothesis survives first.

If it wedges at `8436e362` under WSL too, it is environmental and the tree is exonerated. If it completes, the regression is real and the bisect range is `8436e362..bd075cd5` with the conv-seeding commits already excluded as inert.

**Comment by lukstafi** [comment]
## The diagnosis in this issue is wrong — it is not a middle-end wedge

Re-ran this issue's own repro at `8436e362` under WSL2 on minix (gfx1151, ROCm 7.14), on an idle
machine, with `perf`. The *symptom* reported here is real and reproduces exactly. The diagnosis
does not hold, and two of the three evidence claims are inaccurate.

### 1. 100% of the CPU time is in the HSA runtime, not the compiler

`perf record -F 199 -g` over 20 s of the spinning process, 11,890 samples:

```
66.63%  bench_mlp.exe  libhsa-runtime64.so.1.21.0  rocr::core::Runtime::AsyncEventsLoop(void*)
33.37%  bench_mlp.exe  libhsa-runtime64.so.1.21.0  rocr::core::BusyWaitSignal::WaitRelaxed(...)
```

**Zero OCaml frames.** Re-sampled 85 minutes later: byte-for-byte the same two symbols in the same
2:1 ratio. The process is not lowering, scheduling or digesting anything — it is busy-waiting on an
HSA completion signal for a kernel that has already been compiled and dispatched. The "constant
~3-thread CPU spin" this issue reports is ROCr's busy-wait, not compiler work.

### 2. "Zero debug artifacts written" is not correct

With `--ocannl_output_debug_files_in_build_directory=true`:

```
09:12:53                 process start
09:12:55.5 – 09:12:55.8  12 artifacts, incl. linked .hsaco, for init_params_for_cross_entropy_loss
                         and cross_entropy_loss_forward_and_gradient_then_sgd_update
09:12:55.8 → +2 h        nothing further
```

Codegen and hiprtc are reached and succeed in about 3 seconds. What is zero is artifacts written
*after* the search banner. As phrased ("never reaches rendering or hiprtc, not even for the serial
baseline") the issue sends the reader to the wrong subsystem.

### 3. The mechanism: the serial baseline candidate is unbounded in cost

`Autotune.tune` times the unfissioned serial baseline before any candidate (`time_routine`). The
dispatched kernel `cross_entropy_loss_forward_and_gradient_then_sgd_update.hip` contains **no
`threadIdx` or `blockIdx` reference anywhere** — every loop is a plain serial `for`:

```
for (i219 <= 255) for (i220 <= 1023)                     ->   262 K iterations
for (i95  <= 255) for (i96  <= 1023) for (i97  <= 255)   ->  67.1 M
for (i102 <= 255) for (i103 <= 1023) for (i104 <= 1023)  -> 268.4 M
... plus the backward nests
```

So the entire mlp_wide training step executes in **one work-item** — order 1e9 dependent scalar
iterations — and it executes **four** times, not three: `time_routine` does an untimed warmup run
before the `autotune_repeats=3` loop.

The affected/unaffected table in this issue is then just a per-step-compute ranking:

| cell | step FLOPs (approx) | outcome |
|---|---|---|
| `mlp_small` dims [2,64,64,2] b64 | ~1.7 MFLOP | completes, ~40 s |
| `lenet` | ~30 MFLOP | completes |
| `mlp_wide` dims [256,1024,1024,10] b256 | ~2 GFLOP | "wedges" |
| `cifar_conv` | larger | "wedges" |

Single-lane throughput calibrated from the cells that do complete is 7–46 MFLOP/s; mlp_wide's
1024×1024 weights are cache-resident for nobody on one lane, so it should be lower still. A clean
re-run was **capped at 2 h with no `baseline:` line**, having spent the whole time in
`BusyWaitSignal::WaitRelaxed`. That is the predicted duration for four runs, not a contradiction of
it. The 50-minute watchdog was truncating a search that was progressing.

### 4. It also takes the display driver down

These are multi-hour uninterruptible dispatches on the same device that drives the display. In one
session this produced two Windows-side "AMD Bug Report Tool" driver-timeout notifications, a
transient GPU fault that made a sub-2-minute job hang for 30 minutes (the same job ran in <120 s
afterwards), a silent death of `cifar_stride hip/tuned` at ~11.5 min, and finally **loss of display
output requiring a reboot**. The WSL guest's `dmesg` was clean of GPU faults throughout, so none of
this is visible from inside Linux — which is part of why it kept looking like a hang.

That makes bounding the baseline a stability requirement, not a performance nicety.

### Consequences

- **Retitle**: not "wedges in the middle-end before its first candidate" but "the serial baseline
  candidate is unbounded in cost relative to the workload".
- **The bisect this issue asks for would have chased nothing.** WSL-vs-Windows was never the right
  axis — the symptom reproduces at `8436e362`, and the native-Windows 111.78 s figure is more
  likely a search against a warm `autotune_cache` (the benchmarks README warns it must be wiped for
  `compile_s` to mean anything) than an environment difference.
- **Fix directions**: bound or skip the baseline candidate by predicted cost — the roofline model
  already prices it, and a baseline it prices orders of magnitude off the parallel presets need not
  be measured at all; or seed the baseline from a cheap parallel preset rather than the unscheduled
  form. Either also fixes the driver-stability problem.
- `SKIP_CELLS` entries for `mlp_wide hip/tuned` and the `cifar_conv` watchdog kill are masking a
  cost problem, not a hang.
- **Probably not HIP-specific.** `time_routine` times the serial baseline on every backend, so any
  sufficiently heavy workload should show it. Worth checking the CUDA and Metal legs — the tell is
  a tuned cell that "hangs" where the untuned one is fine.

`benchmarks/report-hip.md` has been corrected accordingly (it carried the same wrong diagnosis).

**Comment by lukstafi** [comment]
### Third affected workload: `cifar_stride hip/tuned`, same profile

Follow-up to the correction above. `cifar_stride` was not in this issue's original table, but it is
the same defect — and it is the cleanest demonstration yet, because the symptom shows up in under a
minute.

Run on the post-gh-527/gh-521 tree, fresh cache, idle machine. A Windows "AMD Bug Report Tool"
driver-timeout notification fired **~45 s after the search started**. Sampled at that moment:

```
66.63%  bench_conv.exe  libhsa-runtime64.so.1.21.0  rocr::core::Runtime::AsyncEventsLoop(void*)
33.35%  bench_conv.exe  libhsa-runtime64.so.1.21.0  rocr::core::BusyWaitSignal::WaitRelaxed(...)
   0    OCaml frames
```

Against `mlp_wide`'s 66.63% / 33.37% / 0 — identical to two decimal places. Same unbounded serial
baseline, so the affected set is `mlp_wide`, `cifar_conv` **and `cifar_stride`**: every workload
above roughly the 30 MFLOP/step mark that `lenet` sits at.

Two things this adds beyond the count:

1. **The 45-second driver timeout is a cheap reproducer.** Diagnosing this does not require the
   multi-hour runs described above — dispatch the baseline on a heavy enough workload and the
   Windows host complains almost immediately, while the guest's `dmesg` stays clean. That is a
   usable acceptance check for whatever bound gets added.
2. **It retires a false lead.** `cifar_stride hip/tuned` completed in 211.61 s at `bd075cd5` and
   then stopped completing, which during this investigation looked like a regression introduced by
   ahrefs/ocannl#484 task 3 (the only library change in that range). It is not — the profile shows
   the ordinary baseline mechanism, and the earlier `bd075cd5` number is best explained by that run
   not having searched from scratch. Anyone bisecting that apparent regression would be chasing
   nothing, same as with the WSL-vs-Windows axis.

Recording this cell as unmeasured-by-this-defect rather than unmeasured-for-unknown-reasons;
`benchmarks/report-hip.md` says so too.

**Comment by lukstafi** [comment]
## Scope after the retitle

The `perf` evidence in lukstafi/ocannl-staging#255 refuted the original diagnosis (100% `libhsa-runtime64`, zero OCaml frames, 12 debug artifacts within 3 s — so neither "middle-end wedge" nor "zero artifacts" held). The title now reflects the mechanism: the autotuner dispatches and times the **serial baseline candidate**, whose kernel has no `threadIdx`/`blockIdx` at all — the whole step in one work-item, run four times (untimed warmup plus `autotune_repeats=3`).

**First step is a 10-minute check, not a fix.** #255 flags "likely not HIP-specific — `time_routine` times the serial baseline on every backend", explicitly unverified on CUDA and Metal. Confirm on Metal before scoping anything: if the serial baseline is timed there too, this is a general autotuner issue and the fix belongs in the timing loop rather than in HIP.

**Then the fix, one of:**

- Don't time an unparallelized candidate on a GPU backend at all — a kernel with no hardware-dimension binding cannot win, so timing it is pure cost. Cheapest, and probably correct.
- Bound candidate execution by wall-clock and treat the overrun as a decline. This is #536's containment problem in a different currency — a time budget rather than an exception — and worth deciding whether #536's typed outcomes should carry it.

**Severity is understated in the original report** and should stay recorded: multi-hour uninterruptible dispatches share the device with the display. One session produced two Windows driver-timeout reports, a transient fault that turned a <2-minute job into a 30-minute hang, and loss of display output requiring a reboot — none of it visible in the guest's `dmesg`. `cifar_stride hip/tuned` has no number in the current report for exactly this reason.

**Comment by lukstafi** [comment]
## Confirmed on gfx1151 (the machine that produced the symptom)

The tuner-side fix (an unparallelized candidate is never dispatched on a GPU backend) is confirmed on HIP/ROCm under WSL, on the same gfx1151 that produced the original hours-long uninterruptible dispatches, Windows-side driver timeouts and lost display. This was the last confirmation the issue was waiting on; Metal was already covered.

All three cells the `SKIP_CELLS` comment named complete, with the display intact and no driver timeout:

| cell | wall | search (`compile_s`) | step p50 |
| --- | --- | --- | --- |
| `mlp_wide` hip/tuned | 34.2 s | 32.1 s | 1.70 ms |
| `cifar_stride` hip/tuned | 131.6 s | 125.3 s | 18.06 ms |
| `cifar_conv` hip/tuned | 181.3 s | 164.6 s | 61.09 ms |

Each was run once, in isolation, under a hard timeout. `baseline_ms` is `inf` on this backend as intended — the serial baseline is never dispatched.

`mlp_wide` was the only actual `SKIP_CELLS` entry; `cifar_conv` and `cifar_stride` were named in its comment as sharing the profile, so all three were measured rather than inferring the other two from one. The entry is dropped, the comment rewritten to record the confirmation, and `test_orchestrate.py` updated (it pinned the entry). The `cifar_conv` metal/tuned entry is a different pathology (reinit-after-tune race) and stays.

One note for anyone using the 45-second figure: that is `mlp_wide`'s scale, not `cifar_stride`'s. `cifar_stride` takes ~2 minutes and `cifar_conv` ~3 on this hardware. All are bounded and interruptible, which is the property that matters here.

Closing from this side — the fix holds where the symptom was born.

**Comment by lukstafi** [comment]
Confirmed on the machine that produced it (lukstafi/ocannl-staging#266). All three cells the `SKIP_CELLS` comment named now complete on gfx1151 under WSL, display intact, no driver timeout:

| cell | wall | search | step p50 |
|---|---|---|---|
| `mlp_wide` hip/tuned | 34.2 s | 32.1 s | 1.70 ms |
| `cifar_stride` hip/tuned | 131.6 s | 125.3 s | 18.06 ms |
| `cifar_conv` hip/tuned | 181.3 s | 164.6 s | 61.09 ms |

All three were measured rather than inferring the other two from `mlp_wide`, which was the only real entry.

**The fix** (lukstafi/ocannl-staging#262): an unparallelized candidate is never dispatched on a GPU backend — `dispatchable` gates the baseline and every candidate, and `baseline_ms` is infinity there, so anything timed outranks it. CPU backends are unchanged, where the serial form runs at full single-core speed and stays a real competitor. A search that timed nothing stores no cache entry and returns the untuned default compile rather than a never-measured incumbent.

**The root cause was general, not HIP-specific**: passing any `?lowered_transform` bypasses the default annotator in `Backend.compile`, so the tuner's base compile was the unscheduled serial form on *every* backend. On Metal that kernel was `grid=[1;1;1] block=[1;1;1] stmts=100` at 6.9 s per run against a 35.7 ms winner, four runs per arm, two arms. Metal's tuned `compile_s` fell 157.2 → 96.3 s with the same winner.

Also caught in review: the cache-hit path returned a saved schedule without consulting the rule, so a pre-fix GPU cache could name the serial baseline as winner — it wins by default whenever every candidate fails to compile, which was exactly the gh-521 state. Fixed by rejecting the replay rather than bumping `entry_version`, which would have discarded every sound entry on every backend.

`cifar_conv` metal/tuned stays in `SKIP_CELLS` — a different pathology, and separately flagged as possibly stale after six non-reproductions.
